### ISSUE TITLE
map files when possible for I/O funcs

### ISSUE THREAD

**Issue by nemequ** [quixdb/squash#120]
Instead of allocating a buffer to compress/decompress to then writing/reading that to the file with fwrite/fread, we should try to use mmap.  Should provide a performance improvement.

**Comment by travisdowns** [comment]
Mmap often makes IO faster, but not always. In particular it depends a lot of the underlying OS implementation of read-ahead. With read() and write() you can explicitly read large chunks, but to get the same effect with mmap the OS has to be smart about it (or you have to use madavise or similar). In my experience recent Linux kernels do a good job. Older Windows versions not so much - only as an example. 

It's also worth noting that switching to memory mapping means that it's tougher to separate the IO from the CPU processing, for example for benchmarking purposes. Mmap always inherently interleaves with the normal operations of compression engine. The ideal access pattern into the input buffer may change with mmap!

Have you considered a mode where the IO happens in another thread, overlapped with the CPU compression work? Perhaps useful for real world use of the squash CLI.

**Comment by nemequ** [comment]
> Mmap often makes IO faster, but not always. In particular it depends a lot of the underlying OS implementation of read-ahead. With read() and write() you can explicitly read large chunks, but to get the same effect with mmap the OS has to be smart about it (or you have to use madavise or similar). In my experience recent Linux kernels do a good job.

Yep, it needs benchmarking.  I actually wrote most of the code for this a few days ago, but it is [#ifdefed out](https://github.com/quixdb/squash/blob/master/squash/file.c#L41) until I have some time to test performance.

> Older Windows versions not so much - only as an example.

It would always be possible to just `#ifdef` out the mapping code on Windows.

> It's also worth noting that switching to memory mapping means that it's tougher to separate the IO from the CPU processing, for example for benchmarking purposes. Mmap always inherently interleaves with the normal operations of compression engine. The ideal access pattern into the input buffer may change with mmap!

Not really; well, at least not for us.  It's not like there are perf counters in Squash which are toggled on and off—the benchmark just [lets the OS handle the details](https://github.com/quixdb/squash-benchmark/blob/master/timer.c).  I/O time doesn't count against CPU time regardless of whether you're using mmap or stdio (except a small amount of time for both when the CPU is actually working, but it's pretty minimal for both).

> Have you considered a mode where the IO happens in another thread, overlapped with the CPU compression work? Perhaps useful for real world use of the squash CLI.

Not really.  I'd be willing to consider it, but spinning up a thread is a pretty big performance hit.  You might see good results for larger files, but it would kill performance for small buffers.  I think you would needs some heuristics for deciding whether or not it is worth it to spin up a thread, and it would only work for regular files (you don't know the size of a pipe, for example).  My gut feeling is that it would probably be better to just mmap reasonably large chunks and let the OS handle shuffling data around.  That's basically the whole idea behind [LMDB](http://symas.com/mdb/), and I can personally vouch for the fact that LMDB works pretty well.

**Comment by travisdowns** [comment]
It could be that recent Windows have improved and it's a win there too. There are a lot of tunables associated with mmap behavior too, even at the OS level, so it's possible that it's a win in some cases and not in others. Will there be options (for CLI and for the library) to control whether it's used or not?

Good point about using CPU time in the benchmark - this keeps IO mostly out of it. I was coming at it mostly from a wall-time angle. Still there are some subtleties there. For example, a read() call generally leaves all the read data in L1 on the core that requested the read (at least for reads small enough to fit in L1), while the behavior of mmap is more varied: it depends on whether the block was already in the page cache, and who put it there, etc. In some cases you find that with mmap wall clock elapsed time comes down, and IO time comes down, but user CPU time goes *up* because the L1 misses happen in user space and get assigned to the process rather than eaten by the kernel while copying the read() result to userspace. Looks like you are including kernel time in timer.c so that shouldn't cause any issue here.

Spinning up a single thread is very quick! You are typically talking 10s or at worst a few 100s of microseconds on modern Linux. Perhaps 10x that on Windows (still fast). In the past, I've measured the method of overlapping IO as pretty much giving exactly the best case theoretical speedup: it completes hides the latency of either the IO, or the [de]compression execution, whichever is faster, and in the case they take around the same time you get about a 2x speedup.

It's hard for the OS to match this, since much of what it does is synchronous, so you don't always get pipelining. In 2.6 kernels, however, you do have async readahead, so when things are working right, it could in principle work as well as an explicit user-space IO thread.

I don't think LMDB is a great comparison actually. Mmap is integral to how LMDB works - unpredictable reads driven by requests, the need to cache blocks, memory pressure driven writeback, etc. The page cache is almost perfect! The IO squash does is a single linear read through a file, which is very predictable and very different.

I'm not actually trying to downplay the benefits of mmap - I actually think it will still be faster in almost all use cases vs the synchronous IO used now (but it's a toss-up vs an explicit IO thread). Just pointing out some of the issues are pretty subtle, and I've seen mmap go wrong before (but a lot of the glass jaws have been fixed by now).

**Comment by nemequ** [comment]
I agree with pretty much everything you wrote there; I definitely want to do a lot of work benchmarking this stuff before I remove the ifdef.

> For example, a read() call generally leaves all the read data in L1 on the core that requested the read (at least for reads small enough to fit in L1)

Yeah, that's a tricky issue.  Chunks that are too small will slow the codecs down (especially if it means they have to do some buffering).  mmap lets us punt the issue to the OS, but I agree that doing so has mixed results.

> Spinning up a single thread is very quick!

Not quick enough.  The Brotli people just rewrote their plugin to switch it from a thread-based streaming implementation to a zlib-style streaming implementation, and there was a *huge* speedup.  According to them, for the smallest files there is a 300-400% difference in performance.

If you're dealing with large files (like most people like to benchmark) the time required to spin up a thread is trivial, but I would argue that there is a bigger market for compressing small pieces of data.  Think about pages from a database (IIRC leveldb compresses 4 KiB blocks [using snappy]), RPC requests/responses, web pages, etc.  A lot of that should be addressable with a thread pool, and a lot of others wouldn't be using I/O (they would just want the buffer-to-buffer API), but it's would still require a lot of careful testing.

Another idea that could be interesting is to use async I/O so we process data as it becomes available instead of letting fread/fwrite block until they have a sufficiently large chunk of data.

> I don't think LMDB is a great comparison actually. Mmap is integral to how LMDB works - unpredictable reads driven by requests, the need to cache blocks, memory pressure driven writeback, etc. The page cache is almost perfect! The IO squash does is a single linear read through a file, which is very predictable and very different.

That's a very good point.

**Comment by nemequ** [comment]
> Will there be options (for CLI and for the library) to control whether it's used or not?

Maybe.  I'd prefer to use some heuristics (either compile- or run-time), but if that can't be done reliably then yes, it would need to be configurable (not sure whether it would be compile- or run-time).

**Comment by travisdowns** [comment]
One approach that I've found works well is to have the heuristic be the default, but have command line options to force one behavior or the other, for expert users, and even to help your own diagnosis of issues. With something with as many caveats and quirks as mmap'd IO, it's very handy to be able to tell some user who is reporting a perf issue "run with this argument to enable/disable feature X and see if that changes things".

**Comment by nemequ** [comment]
Yes, I basically did that for splicing with d9b82bd8a5062e5f11f115276e1e3c31c004b2d7, I'll probably do something similar for this.
